src/mappings/poktroll/gateways.ts

```diff
@@ -0,0 +1,137 @@
+import { CosmosEvent, CosmosMessage } from "@subql/types-cosmos";
+import {
+  Gateway,
+  MsgStakeGateway as MsgStakeGatewayEntity,
+  MsgUnstakeGateway as MsgUnstakeGatewayEntity,
+  EventGatewayUnstaked as EventGatewayUnstakedEntity,
+} from "../../types";
+import { MsgStakeGateway, MsgUnstakeGateway } from "../../types/proto-interfaces/poktroll/gateway/tx";
+import { StakeStatus } from "../constants";
+import {
+  attemptHandling,
+  getEventId,
+  messageId,
+  stringify,
+  unprocessedEventHandler,
+  unprocessedMsgHandler,
+} from "../utils";
+
+export async function handleGatewayMsgStake(msg: CosmosMessage<MsgStakeGateway>): Promise<void> {
+  await attemptHandling(msg, _handleGatewayMsgStake, unprocessedMsgHandler)
+}
+
+export async function handleGatewayMsgUnstake(msg: CosmosMessage<MsgUnstakeGateway>): Promise<void> {
+  await attemptHandling(msg, _handleGatewayMsgUnstake, unprocessedMsgHandler)
+}
+
+export async function handleGatewayUnstakeEvent(event: CosmosEvent): Promise<void> {
+  await attemptHandling(event, _handleGatewayUnstakeEvent, unprocessedEventHandler);
+}
+
+async function _handleGatewayMsgStake(
+  msg: CosmosMessage<MsgStakeGateway>
+) {
+  logger.info(`[handleGatewayMsgStake] (msg.msg): ${stringify(msg.msg, undefined, 2)}`);
+
+  if (!msg.msg.decodedMsg.stake) {
+    throw new Error(`[handleGatewayMsgStake] stake not provided in msg`)
+  }
+
+  const stake = {
+    amount: msg.msg.decodedMsg.stake.amount,
+    denom: msg.msg.decodedMsg.stake.denom,
+  }
+
+  const gateway = Gateway.create({
```

Comment:
Same feedback as suppliers and applications; I think we need to check for an existing gateway and reuse it if one exists before creating a new one.